src/dawn/IIR/IIRSerializer.cpp

```diff
@@ -0,0 +1,833 @@
+}
+
+static void
+serializeStmtAccessPair(proto::iir::StatementAccessPair* protoStmtAccessPair,
+                        const std::unique_ptr<iir::StatementAccessesPair>& stmtAccessPair) {
+  // serialize the statement
+  ProtoStmtBuilder builder(protoStmtAccessPair->mutable_statement()->mutable_aststmt());
+  stmtAccessPair->getStatement()->ASTStmt->accept(builder);
+
+  // check if caller / callee accesses are initialized, and if so, fill them
+  if(stmtAccessPair->getCallerAccesses()) {
+    setAccesses(protoStmtAccessPair->mutable_calleraccesses(), stmtAccessPair->getCallerAccesses());
+  }
+  if(stmtAccessPair->getCalleeAccesses()) {
+    setAccesses(protoStmtAccessPair->mutable_calleeaccesses(), stmtAccessPair->getCalleeAccesses());
+  }
+}
+
+static void setCache(proto::iir::Cache* protoCache, const iir::Cache& cache) {
+  protoCache->set_accessid(cache.getCachedFieldAccessID());
+  switch(cache.getCacheIOPolicy()) {
+  case iir::Cache::bpfill:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_BPFill);
+    break;
+  case iir::Cache::epflush:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_EPFlush);
+    break;
+  case iir::Cache::fill:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_Fill);
+    break;
+  case iir::Cache::fill_and_flush:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_FillFlush);
+    break;
+  case iir::Cache::flush:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_Flush);
+    break;
+  case iir::Cache::local:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_Local);
+    break;
+  case iir::Cache::unknown:
+    protoCache->set_policy(proto::iir::Cache_CachePolicy_CP_Unknown);
+    break;
+  default:
+    dawn_unreachable("unknown cache policy");
+  }
+  switch(cache.getCacheType()) {
+  case iir::Cache::bypass:
+    protoCache->set_type(proto::iir::Cache_CacheType_CT_Bypass);
+    break;
+  case iir::Cache::IJ:
+    protoCache->set_type(proto::iir::Cache_CacheType_CT_IJ);
+    break;
+  case iir::Cache::IJK:
+    protoCache->set_type(proto::iir::Cache_CacheType_CT_IJK);
+    break;
+  case iir::Cache::K:
+    protoCache->set_type(proto::iir::Cache_CacheType_CT_K);
+    break;
+  default:
+    dawn_unreachable("unknown cache type");
+  }
+  if(cache.getInterval().is_initialized()) {
+    auto sirInterval = cache.getInterval()->asSIRInterval();
+    setInterval(protoCache->mutable_interval(), &sirInterval);
+  }
+  if(cache.getEnclosingAccessedInterval().is_initialized()) {
+    auto sirInterval = cache.getEnclosingAccessedInterval()->asSIRInterval();
+    setInterval(protoCache->mutable_enclosingaccessinterval(), &sirInterval);
+  }
+  if(cache.getWindow().is_initialized()) {
+    protoCache->mutable_cachewindow()->set_minus(cache.getWindow()->m_m);
+    protoCache->mutable_cachewindow()->set_plus(cache.getWindow()->m_p);
+  }
+}
+
+static iir::Cache makeCache(const proto::iir::Cache* protoCache) {
```

Comment:
above you use `setCache` and here `makeCache` and with different signature. I would try to be consistent, and name it after serialization, since we are not making and object but simply deserializing. 

```
static void serializeCache(proto::iir::Cache& protoCache, const iir::Cache& cache) ;
static void deserializeCache(iir::Cache& cache, const proto::iir::Cache& protoCache) ;
```

Reply:
this is just to keep it consistent with how we did it for everything (see ASTSerializer for examples) so far:

`setXXX` sets the proto instantiation of `XXX` from the constant object in memory. This is a void return as otherwise the mutable proto-objects are a pain to handle
`makeXXX` returns the in-memory representation of the concept XXX.